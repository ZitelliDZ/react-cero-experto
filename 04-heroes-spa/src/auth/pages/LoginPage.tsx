import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../context/AuthContext";



export const LoginPage = () => {

    const navigate = useNavigate();
    const { login } = useContext(AuthContext);

    const onLogin = () => {
      const lastPath = localStorage.getItem("lastPath") || "/";
      login('Diego');
      navigate( lastPath
        , { replace: true });
    }

    const onSubmit = (e: React.FormEvent<HTMLFormElement>) => {
      e.preventDefault();
     
    }
    return (
      <div className="container mt-5">
        <h1>Login</h1>
        <hr />
        <div className="row">
          <div className="col-md-6">
            <form  onSubmit={onSubmit}>
              <div className="mb-3">
                <label className="form-label">Email address</label>
                <input type="email" className="form-control" />
              </div>
              <div className="mb-3">
                <label className="form-label">Password</label>
                <input type="password" className="form-control" />
              </div>
              <button type="submit" className="btn btn-primary" onClick={onLogin}>
                Submit
              </button>
            </form>
          </div>
        </div>
        
      </div>
    )
  }