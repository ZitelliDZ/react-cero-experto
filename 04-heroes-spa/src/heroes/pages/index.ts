export { MarvelPage } from "./MarvelPage";
export { DcPage } from "./DcPage";
export { HeroPage } from "./HeroPage";
export { SearchPage } from "./SearchPage";