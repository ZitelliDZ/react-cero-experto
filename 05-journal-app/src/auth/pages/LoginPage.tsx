import { Google } from "@mui/icons-material";
import { Link as RouterLink } from "react-router-dom";
import { Alert, Button, Grid, Link, TextField, Typography } from "@mui/material"; 
import { AuthLayout } from "../layout/AuthLayout";
import { useForm } from "../../hook/useForm";
import { startGoogleSignIn, startLoginWithEmailAndPassword } from "../../store/auth/thunks";
import { useDispatch, useSelector } from "react-redux";  
import { useMemo } from "react";

const initialFormState = {
  email: "",
  password: "",
}

export const LoginPage = () => {


  const { status,errorMessage } = useSelector((state) => state.auth);

  const isAutencticated = useMemo(() => status === "checking", [status]); 

  const dispatch = useDispatch();

  const { formState, onInputChange, onResetForm } = useForm( initialFormState );

  const { email, password } = formState;

  const onSubmit = (e) => {
    e.preventDefault();
    dispatch(startLoginWithEmailAndPassword(email, password));
    onResetForm();
  };

  const onGoogleLogin = async() => {
    console.log('onGoogleSignIn');
    dispatch( startGoogleSignIn() ); 
  }

  return (
    <AuthLayout title="Login">
      <form onSubmit={onSubmit}
      
      className="animate__animated animate__fadeIn animate__faster"
      >
        <Grid container spacing={2} xs={12}> 
          <Grid item xs={12}>
            <TextField
              label="Email"
              type="email"
              placeholder="email@gmail.com"
              className="input"
              fullWidth
              name="email"
              value={email}
              onChange={onInputChange}
            />
          </Grid>
          <Grid item xs={12}>
            <TextField
              label="ContraseÃ±a"
              type="password"
              placeholder="*********"
              className="input"
              fullWidth
              name="password"
              value={password}
              onChange={onInputChange}
            />
          </Grid>
          <Grid
            container
            xs={12}
            spacing={2}
            sx={{ mb: 2, mt: 2 }}
            style={{ marginLeft: "0px" }}
          >
          <Alert severity="error" xs={12} sx={{ display: !!errorMessage ? "" : "none" }}>
            {errorMessage}
          </Alert>
            <Grid item xs={12} md={6} lg={6}>
              <Button 
              disabled={isAutencticated}
              
              variant="contained" color="primary" fullWidth type="submit">
                Login
              </Button>
            </Grid>
            <Grid item xs={12} md={6} lg={6}>
              <Button variant="contained" color="primary" fullWidth onClick={onGoogleLogin}
              disabled={isAutencticated}
              >
                <Google />
                <Typography sx={{ ml: 1 }}>Google</Typography>
              </Button>
            </Grid>
          </Grid>

          <Grid container direction={"row"} justifyContent={"end"} xs={12}>
            <Link to="/auth/register" color={"inherit"} component={RouterLink}>
              Crear cuenta
            </Link>
          </Grid>
        </Grid>
      </form>
    </AuthLayout>
  );
};
