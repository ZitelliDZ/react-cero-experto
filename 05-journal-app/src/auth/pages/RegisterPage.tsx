import { Alert, Button, Grid, Link, TextField, Typography } from "@mui/material"
import { AuthLayout } from "../layout/AuthLayout"
import { Google } from "@mui/icons-material"
import { Link as LinkRouter } from "react-router-dom"
import { useForm } from "../../hook/useForm"
import { useMemo, useState } from "react"
import { startCreatingUserWithEmailAndPassword } from "../../store/auth/thunks"
import { useDispatch, useSelector } from "react-redux"

const formValidations = {
  email: [(value)=>value.includes('@'), 'Email no válido'],
  password: [(value)=>value.length > 5, 'La contraseña debe tener al menos 6 caracteres'],

  name: [(value)=>value.length > 0, 'El nombre es requerido'],
}


export const RegisterPage = () => {

  const [ formSubmitted, setFormSubmitted ] = useState(false)

  const dispatch = useDispatch()
  const { status, errorMessage } = useSelector((state) => state.auth)
  const isCheckingAuthentication = useMemo(() => status === "checking", [status]) 

  const { formState, onInputChange, onResetForm,formValidationsState,isFormValid } = useForm({
    name: "",
    email: "",
    password: "",
  },formValidations)
 

  const onSubmit = (e) => {
    e.preventDefault()
    setFormSubmitted(true)
    if(!isFormValid) return

    
    dispatch(startCreatingUserWithEmailAndPassword(formState.email, formState.password, formState.name))
    onResetForm()
  }

  return (
    <AuthLayout title="Registro">
       
      <form onSubmit={onSubmit}
      
      className="animate__animated animate__fadeIn animate__faster"
      >
        <Grid container spacing={2} xs={12}>
          <Grid item xs={12}>
            <TextField
              label="Nombre Completo"
              type="text"
              placeholder="Jhon Doe"
              className="input"
              fullWidth
              name="name"
              value={formState.name}
              onChange={onInputChange}

              error={!!formValidationsState.nameValid && formSubmitted}
              helperText={formValidationsState.nameValid}


            />
          </Grid>
          <Grid item xs={12}>
            <TextField
              label="Email"
              type="email"
              placeholder="email@gmail.com"
              className="input"
              fullWidth
              name="email"
              value={formState.email}
              onChange={onInputChange}
              error={!!formValidationsState.emailValid && formSubmitted}
              helperText={formValidationsState.emailValid}
            />
          </Grid>
          <Grid item xs={12}>
            <TextField
              label="Contraseña"
              type="password"
              placeholder="*********"
              className="input"
              fullWidth
              name="password"
              value={formState.password}
              onChange={onInputChange}
              error={!!formValidationsState.passwordValid && formSubmitted}
              helperText={formValidationsState.passwordValid}
            />
          </Grid>
          <Grid
            container
            xs={12}
            spacing={2}
            sx={{ mb: 2, mt: 2 }}
            style={{ marginLeft: "0px" }}
          >
            <Alert severity="error" xs={12} sx={{ display: !!errorMessage ? "" : "none" }}>
              {errorMessage}
            </Alert>
            <Grid item xs={12} md={6} lg={6}>
              <Button variant="contained" color="primary" fullWidth type="submit"   disabled={isCheckingAuthentication}>
                Crear
              </Button>
            </Grid> 
          </Grid>

          <Grid container direction={"row"} justifyContent={"end"} xs={12}>
            <Typography sx={{ mr:1 }}>¿Ya tienes una cuenta?</Typography>
            <Link to="/auth/login" color={"inherit"} component={LinkRouter}>
              Iniciar sesión
            </Link>
          </Grid>
        </Grid>
      </form>
    </AuthLayout>
  )
}
