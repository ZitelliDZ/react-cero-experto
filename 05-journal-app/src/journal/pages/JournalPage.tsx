 
import { IconButton } from "@mui/material"
import JournalLayout from "../layout/JournalLayout"
import { NoteView } from "../views/NoteView"
import { NothingSelectedView } from "../views/NothingSelectedView"
import { AddOutlined } from "@mui/icons-material"
import { startNewNote } from "../../store/journal/thunks"
import { useDispatch, useSelector } from "react-redux"



export const JournalPage = () => {

  const dispatch = useDispatch()

  const { isSaving, active } = useSelector((state) => state.journal)

  const onClickNewNote = () => {
    dispatch(startNewNote())
  }
    
  return (
    <JournalLayout>
 

        {
          (!!active) 
          ? <NoteView />
          : <NothingSelectedView />
        } 

        <IconButton 
        onClick={onClickNewNote}
        size="large"
        disabled={isSaving}
        sx={{ 
          color: 'white',

          ':disabled': {
            backgroundColor: 'error.light',
            opacity: [0.9, 0.8, 0.7],
          },

          backgroundColor: 'error.main',
          ':hover': {
            backgroundColor: 'error.dark',
            opacity: [0.9, 0.8, 0.7],
          },
          position: 'fixed',
          right: 50,
          bottom: 50,
         }}
         >
          <AddOutlined sx={{ fontSize: 30 }} />

         </IconButton>

        

    </JournalLayout>
  )
}
