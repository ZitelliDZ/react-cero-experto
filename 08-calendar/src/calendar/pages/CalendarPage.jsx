import { Calendar } from "react-big-calendar";
import "react-big-calendar/lib/css/react-big-calendar.css";

import NavBar from "../components/NavBar";
import { addHours } from "date-fns";
import { localizer } from "../../helpers/calendarLocalizer";
import { getMessagesEsp } from "../../helpers/calendarCulture";
import CalendarEventBox from "../components/CalendarEventBox";
import { useState } from "react";
import { CalendarModal } from "../components/CalendarModal";
import { useUiStore } from "../../hooks/useUiStore";
import { useCalendarStore } from "../../hooks/useCalendarStore";
import { FabNewButton } from "../components/FabNewButton";
import { FabDeleteButton } from "../components/FabDeleteButton";

 

export const CalendarPage = () => {
  const [lastView, setLastView] = useState(
    localStorage.getItem("lastView") || "month"
  );

  const { events,setActiveEvent,hasEventSelected } = useCalendarStore();

  const { openDateModal } = useUiStore();

  const eventStyleGetter = (event, start, end, isSelected) => {
    const style = {
      backgroundColor: "#367CF7",
      borderRadius: "5px",
      opacity: 0.8,
      display: "block",
      color: "white",
    };
    return {
      style,
    };
  };

  const onDoubleClick = (e) => {
    openDateModal();
  };
  const onSelectEvent = (e) => {
    setActiveEvent(e);
 
  };
  const onViewChanged = (e) => {};

  return (
    <>
      <NavBar />
      <Calendar
        culture="es"
        localizer={localizer}
        defaultView={lastView}
        events={events}
        startAccessor="start"
        endAccessor="end"
        style={{ height: "calc(100vh - 80px)" }}
        messages={getMessagesEsp()}
        eventPropGetter={eventStyleGetter}
        components={{
          event: CalendarEventBox,
        }}
        onDoubleClickEvent={onDoubleClick}
        onSelectEvent={onSelectEvent}
        onView={onViewChanged}
      />
      <CalendarModal />
      <FabNewButton />
      {hasEventSelected && <FabDeleteButton />}

    </>
  );
};
