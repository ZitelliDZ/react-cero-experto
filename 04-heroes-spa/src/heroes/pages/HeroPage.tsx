import { useEffect, useMemo, useState } from "react";
import { useNavigate, useParams } from "react-router-dom";
import { Hero } from "../data/heroes";
import { getHeroeById } from "../helpers/getHeroeById";
import { HeroCard } from "../components/HeroCard";

export const HeroPage = () => {
  const { heroId } = useParams();
  const navigate = useNavigate();

  const [hero, setHero] = useState<Hero | null>(null);

  const heroResult = useMemo(() => getHeroeById(heroId!),[heroId]);

  useEffect(() => {
    if (heroId) {
      try {
        setHero(heroResult);
      } catch (error) {
        console.error(error);
      }
    }
  }, [heroId]);

  if (!hero) {
    return <div>Hero not found</div>;
    //return <Navigate to="/marvel" />;
  }

  const onReturn = () => {
    navigate(-1);
  }

  return (
    <div>
      <h1>Hero Page</h1>
      <hr />
      <HeroCard hero={hero}  />
      <button className="btn btn-primary mt-4" onClick={ onReturn }>
        Regresar
      </button>
    </div>
  );
};
