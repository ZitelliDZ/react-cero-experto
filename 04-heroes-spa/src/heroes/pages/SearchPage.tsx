import { useLocation, useNavigate } from "react-router-dom";
import { HeroCard } from "../components/HeroCard";
import { useForm } from "../hooks/useForm";
import queryString from "query-string";
import { useEffect, useState } from "react";
import { Hero } from "../data/heroes";
import { getHeroByName } from "../helpers/getHeroByName";

export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const [heros, setHeros] = useState<Hero[] | null>(null);

  const { q = "" } = queryString.parse(location.search);

  const { formState, onInputChange } = useForm({
    search: q,
  });

  useEffect(() => { 
      try {
        const heroResult = getHeroByName(q);
        setHeros(heroResult);
      } catch (error) {
        setHeros(null);
        console.error(error);
      } 
  }, [q ]);
  const handleSubmit = (e: any) => {
    e.preventDefault();
    navigate(`?q=${formState.search}`);
  };

  return (
    <>
      <h1>Search Page</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Search Form</h4>
          <hr />
          <form onSubmit={handleSubmit}>
            <input
              type="text"
              placeholder="Find your hero"
              className="form-control"
              name="search"
              value={formState.search}
              onChange={onInputChange}
            />
            <button
              type="submit"
              className="btn m-1 mt-3 btn-block btn-outline-primary"
            >
              Search...
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {/* {!heros &&
            (!q ? (
              <div className="alert alert-info">Search a hero</div>
            ) : (
              <div className="alert alert-danger">Heroe {q} no encontrado</div>
            ))} */}


            <div className="alert alert-info animate__animated animate__fadeIn" aria-label="alerta-info"
            style={{display: q !== '' ? "none" : " "}}>
              Search a hero</div>

            <div className="alert alert-danger animate__animated animate__fadeIn" aria-label="alerta-danger" 
            style={{display: (q !== '' && !heros || heros?.length == 0) ? "" : "none"}}>
              Heroe {q} no encontrado</div>

               

          {heros && (
            <div className="d-flex w-100 flex-row flex-wrap gap-4 animate__animated animate__fadeIn">
              {heros.map((hero) => ( 
                  <HeroCard  hero={hero}  key={hero.id} />
              ))}
            </div>
          )}
        </div>
      </div>
    </>
  );
};
