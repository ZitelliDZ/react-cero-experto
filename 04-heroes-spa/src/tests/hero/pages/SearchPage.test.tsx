import { render, screen } from "@testing-library/react";
import { MemoryRouter, useNavigate } from "react-router-dom";
import { SearchPage } from "../../../heroes/pages";

// const mockUseNavigate = jest.fn();
// jest.mock("react-router-dom", () => ({
//     ...jest.requireActual("react-router-dom"),
//     useNavigate:  mockUseNavigate,
// }));


describe('SearchPage', () => {

    // beforeEach(() => {
    //     jest.clearAllMocks();
    // });

    test('debe de mostrarse correctamente', () => {
        const {container} = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        )

        screen.debug();
        expect(container).toMatchSnapshot();

    });

    test('debe de mostrar a Batman y el input con el valor del queryString', () => {

        const {container} = render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        )

        screen.debug();
        expect(screen.getByDisplayValue('batman')).toBeTruthy();
        expect(screen.getByText('Batman')).toBeTruthy();
        //expect(container).toMatchSnapshot();

        const input = screen.getByRole('textbox');
        expect(input.value).toBe('batman'); 
        const img = screen.getByRole('img');
        expect(img.src).toContain('/assets/heroes/dc-batman.jpg');

        const alertInfo = screen.getByLabelText('alerta-info');
        expect(alertInfo.style.display).toBe('none');
        const alertDanger = screen.getByLabelText('alerta-danger');
        expect(alertDanger.style.display).toBe('none');
    });

    test('debe de mostrar un error si el hero no existe', () => {

        const q = 'batman123';
        const {container} = render(
            <MemoryRouter initialEntries={['/search?q='+q]}>
                <SearchPage />
            </MemoryRouter>
        )

        screen.debug();
        expect(screen.getByText('Heroe '+q+' no encontrado')).toBeTruthy();
        //expect(container).toMatchSnapshot();

        const alertInfo = screen.getByLabelText('alerta-info');
        expect(alertInfo.style.display).toBe('none');
        const alertDanger = screen.getByLabelText('alerta-danger');
        expect(alertDanger.style.display).toBe('');
    });

    test('debe de mostrar un error si el queryString está vacío', () => {

        const {container} = render(
            <MemoryRouter initialEntries={['/search?q=']}>
                <SearchPage />
            </MemoryRouter>
        )

        screen.debug();
        expect(screen.getByText('Search a hero')).toBeTruthy();
        //expect(container).toMatchSnapshot();

        const alertInfo = screen.getByLabelText('alerta-info');
        expect(alertInfo.style.display).toBe('');
        const alertDanger = screen.getByLabelText('alerta-danger');
        expect(alertDanger.style.display).toBe('none');
    });
});