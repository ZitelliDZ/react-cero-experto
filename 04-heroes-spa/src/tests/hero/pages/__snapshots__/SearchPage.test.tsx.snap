// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    Search Page
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form>
        <input
          class="form-control"
          name="search"
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          class="btn m-1 mt-3 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="alerta-info"
        class="alert alert-info animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
      <div
        aria-label="alerta-danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        Heroe 
         no encontrado
      </div>
    </div>
  </div>
</div>
`;
